backends/ebpf/runtime/ebpf_runtime.c

```diff
@@ -14,89 +14,270 @@ See the License for the specific language governing permissions and
     fprintf(stderr, "\t-f: The input pcap file\n");
+    fprintf(stderr, "\t-n: Specifies the number of input files\n");
     exit(EXIT_FAILURE);
 }
 
-static int process_packets(pcap_t *in_handle, pcap_dumper_t *out_handle) {
-    struct pcap_pkthdr *pcap_hdr;
-    const unsigned char *packet;
-    int ret;
-    while ((ret = pcap_next_ex(in_handle, &pcap_hdr, &packet)) == 1) {
+/* Struct containing packet content and its metadata.
+ For now this is only the pcap header and the input interface.
+ */
+typedef struct {
+    char *data;
+    struct pcap_pkthdr pcap_hdr;
+    uint32_t ifindex;
+} pcap_pkt;
+
+/* Carries a list of packets and the total length as information.
+ Used as input stream for the packet parsing interface.
+ */
+typedef struct {
+    pcap_pkt *list;
+    uint64_t len;
+} pcap_pkt_list;
+
+/* Rank packets based on the timestamp of the pcap header */
+int compare(const void *s1, const void *s2) {
+  pcap_pkt *p1 = (pcap_pkt *)s1;
+  pcap_pkt *p2 = (pcap_pkt *)s2;
+  return p1->pcap_hdr.ts.tv_usec > p2->pcap_hdr.ts.tv_usec;
+}
+
+static pcap_pkt_list *sort_packets(pcap_pkt_list *pkt_lists, int num_lists) {
+    /* Get the total length of all individual lists combined first */
+    uint64_t total_len = 0;
+    for (int i = 0; i < num_lists; i++)
+        total_len+= pkt_lists[i].len;
+    /* Allocate the "master" list */
+    pcap_pkt_list *sorted_list = malloc(sizeof(pcap_pkt_list));
+    pcap_pkt *full_pkts = malloc(sizeof(pcap_pkt) * total_len);
+
+    /* Fill the master list by copying over the individual packets */
+    uint64_t offset = 0;
+    for (int i = 0; i < num_lists; i++) {
+        memcpy(&full_pkts[offset], pkt_lists[i].list, sizeof(pcap_pkt) *pkt_lists[i].len);
+        offset += pkt_lists[i].len;
+        /* We do not need the temporary allocated list anymore */
+        free(pkt_lists[i].list);
+    }
+    /* Sort the master list and return the meta information*/
+    qsort(full_pkts, total_len, sizeof(pcap_pkt), compare);
+    sorted_list->list = full_pkts;
+    sorted_list->len = total_len;
+    return sorted_list;
+}
+
+static void fill_packets(pcap_pkt_list *pkt_list, int num_files, pcap_t **in_handles) {
+    /* Extract all pcap packets per output file */
+    for (int i = 0; i < num_files; i++) {
+        struct pcap_pkthdr *pcap_hdr;
+        const unsigned char *packet;
+        char errbuf[PCAP_ERRBUF_SIZE];
+        int ret;
+        FILE *pcap_in_fp = pcap_file(in_handles[i]);
+        uint64_t pkt_index = 0;
+        /* Count packets in the pcap file for allocation*/
+        while ((ret = pcap_next_ex(in_handles[i], &pcap_hdr, &packet)) == 1)
+            pkt_index++;
+        if (ret == -1) {
+            pcap_perror(in_handles[i], "Error: Failed to parse ");
+            exit(EXIT_FAILURE);
+        }
+        /* Reset the file pointer and reopen the pcap handle */
+        fseek(pcap_in_fp, 0, 0);
+        in_handles[i] = pcap_fopen_offline(pcap_in_fp, errbuf);
```

Comment:
You haven't closed the file. Why do you need to reopen?
Why don't you just open offline from the start?